### PR TITLE
arm64/a64: Add driver for MIPI DSI

### PR DESCRIPTION
**Pull request by lupyuen**
(Squashed the commits from https://github.com/apache/nuttx/pull/7846)

## Summary

This PR adds the driver for Allwinner A64's MIPI Display Serial Interface (DSI) and MIPI Display Physical Layer (D-PHY).

This driver will be used by the upcoming Display Driver for PINE64 PinePhone.

### Modified Files

- `include/nuttx/crc16.h`: Added 16-bit CRC-CCITT

- `libs/libc/misc/Make.defs`: Added 16-bit CRC-CCITT to Makefile

- `arch/arm64/src/a64/Kconfig`: Added the Kconfig option for "A64 Peripheral Selection > MIPI DSI" (`CONFIG_A64_MIPI_DSI`), which enables the MIPI DSI Driver

- `arch/arm64/src/a64/hardware/a64_memorymap.h`: Added the Base Address for MIPI DSI

- `arch/arm64/src/a64/Make.defs`: Added the MIPI DSI Driver to the Makefile

### New Files

- `libs/libc/misc/lib_crc16ccitt.c`: Compute 16-bit CRC-CCITT

- `arch/arm64/src/a64/mipi_dsi.c`, `mipi_dsi.h`: Compose MIPI DSI Packets (Long, Short, Short with Parameter)

- `arch/arm64/src/a64/a64_mipi_dsi.c`, `a64_mipi_dsi.h`: MIPI DSI Driver for Allwinner A64

- `arch/arm64/src/a64/a64_mipi_dphy.c`, `a64_mipi_dphy.h`: MIPI D-PHY Driver for Allwinner A64

### Updated Documentation

- `platforms/arm/a64/boards/pinephone/index.rst`: Added MIPI DSI as supported peripheral for PinePhone

## Impact

Our MIPI DSI Driver will be built only when we select the Kconfig option for "A64 Peripheral Selection > MIPI DSI" (`CONFIG_A64_MIPI_DSI`).

The MIPI DSI Driver will be called by the upcoming drivers for A64 Timing Controller (TCON) and A64 Display Engine.

Our MIPI DSI Driver has no impact on existing code because it's not called yet by the existing code.

The Driver Code is derived from my Reverse-Engineering of the PinePhone Display Driver, based on the logs captured from the PinePhone p-boot Bootloader:

-   ["Understanding PinePhone's Display (MIPI DSI)"](https://lupyuen.github.io/articles/dsi)

-   ["NuttX RTOS for PinePhone: Display Driver in Zig"](https://lupyuen.github.io/articles/dsi2)

## Testing

Some parts of the PinePhone Display Driver are still under construction. To test the MIPI DSI Driver, we ran a Zig Program that integrates the NuttX MIPI DSI Driver with the upcoming parts of the Display Driver:

-  ["Test MIPI DSI for NuttX Kernel"](https://github.com/lupyuen/pinephone-nuttx#test-mipi-dsi-for-nuttx-kernel)

The Integration Test [successfully rendered graphics](https://lupyuen.github.io/images/dsi3-title.jpg) on the LCD Display. Here's the Test Log:

-  [Test Log for NuttX MIPI DSI on PinePhone](https://gist.github.com/lupyuen/f1a02068aeb0785278c482116a4eedc7)

We also ran Unit Tests for the MIPI DSI Driver, [as explained here.](https://github.com/lupyuen/pinephone-nuttx#test-mipi-dsi-for-nuttx-kernel)

### CODE REVIEW

File: arch/arm64/src/a64/mipi_dsi.c [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+#include <debug.h>
+#include "mipi_dsi.h"
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* From CRC-16-CCITT (x^16 + x^12 + x^5 + 1) */
+
+static uint16_t crc16ccitt_tab[256] =
```
[the remark this answers is not in the document]
Comment (lupyuen) [reply arch/arm64/src/a64/mipi_dsi.c]: CRC-CCITT 16-bit has been moved to libc, MIPI DSI Driver tested OK. I have rebased with master. Thanks :-)

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm64/src/a64/a64_mipi_dphy.c [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* A64 CCU Registers and Bit Definitions ************************************/
+
+/* MIPI_DSI Clock Register */
+
+#define MIPI_DSI_CLK_REG    (A64_CCU_ADDR + 0x168)
+#define DPHY_CLK_DIV_M(n)   (n << 0)
+#define DSI_DPHY_SRC_SEL(n) (n << 8)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dphy.c:57]:
```suggestion
#define DPHY_CLK_DIV_M(n)   ((n) << 0)
#define DSI_DPHY_SRC_SEL(n) ((n) << 8)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 3]
```diff
@@ -0,0 +1,993 @@
...
+#define ECC_EN             (1 << 16)
+#define CRC_EN             (1 << 17)
+
+/* DSI Configuration Register 1 (A31 Page 846) */
+
+#define DSI_BASIC_CTL1_REG (A64_DSI_ADDR + 0x14)
+#define DSI_MODE                   (1 << 0)
+#define VIDEO_FRAME_START          (1 << 1)
+#define VIDEO_PRECISION_MODE_ALIGN (1 << 2)
+#define VIDEO_START_DELAY(n)       (n << 4)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:99]:
```suggestion
#define VIDEO_START_DELAY(n)       ((n) << 4)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 4]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Line Number Register 0 (A31 Page 847) */
+
+#define DSI_BASIC_SIZE0_REG (A64_DSI_ADDR + 0x18)
+#define VIDEO_VSA(n)        (n << 0)
+#define VIDEO_VBP(n)        (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:105]:
```suggestion
#define VIDEO_VSA(n)        ((n) << 0)
#define VIDEO_VBP(n)        ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 5]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Line Number Register 1 (A31 Page 847) */
+
+#define DSI_BASIC_SIZE1_REG (A64_DSI_ADDR + 0x1c)
+#define VIDEO_VACT(n)       (n << 0)
+#define VIDEO_VT(n)         (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:111]:
```suggestion
#define VIDEO_VACT(n)       ((n) << 0)
#define VIDEO_VT(n)         ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 6]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Instruction Function Register (Undocumented) */
+
+#define DSI_INST_FUNC_REG(n)   (A64_DSI_ADDR + 0x020 + n * 0x04)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:115]:
```suggestion
#define DSI_INST_FUNC_REG(n)   (A64_DSI_ADDR + 0x020 + (n) * 0x04)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 7]
```diff
@@ -0,0 +1,993 @@
...
+#define DSI_INST_FUNC_LANE_CEN (1 << 4)
+
+/* DSI Instruction Loop Select Register (Undocumented) */
+
+#define DSI_INST_LOOP_SEL_REG  (A64_DSI_ADDR + 0x40)
+
+/* DSI Instruction Loop Number Register (Undocumented) */
+
+#define DSI_INST_LOOP_NUM_REG(n) (A64_DSI_ADDR + 0x44 + n * 0x10)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:124]:
```suggestion
#define DSI_INST_LOOP_NUM_REG(n) (A64_DSI_ADDR + 0x44 + (n) * 0x10)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 8]
```diff
@@ -0,0 +1,993 @@
...
+#define DSI_INST_ID_HSD       3
+#define DSI_INST_ID_LPDT      4
+#define DSI_INST_ID_HSCEXIT   5
+#define DSI_INST_ID_NOP       6
+#define DSI_INST_ID_DLY       7
+#define DSI_INST_ID_END       15
+
+/* DSI Instruction Jump Configuration Register (Undocumented) */
+
+#define DSI_INST_JUMP_CFG_REG(n) (A64_DSI_ADDR + 0x4c + n * 0x04)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:141]:
```suggestion
#define DSI_INST_JUMP_CFG_REG(n) (A64_DSI_ADDR + 0x4c + (n) * 0x04)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 9]
```diff
@@ -0,0 +1,993 @@
...
+#define DSI_TRANS_ZERO_REG  (A64_DSI_ADDR + 0x78)
+
+/* DSI Timing Controller DRQ Register (Undocumented) */
+
+#define DSI_TCON_DRQ_REG    (A64_DSI_ADDR + 0x7c)
+
+/* DSI Pixel Format Register 0 (A31 Page 847) */
+
+#define DSI_PIXEL_CTL0_REG (A64_DSI_ADDR + 0x80)
+#define PIXEL_FORMAT(n) (n << 0)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:159]:
```suggestion
#define PIXEL_FORMAT(n) ((n) << 0)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 10]
```diff
@@ -0,0 +1,993 @@
...
+#define PIXEL_ENDIAN    (0 << 4)
+#define PD_PLUG_DIS     (1 << 16)
+
+/* DSI Pixel Package Register 0 (A31 Page 848) */
+
+#define DSI_PIXEL_PH_REG (A64_DSI_ADDR + 0x90)
+#define PIXEL_DT(n)      (n << 0)
+#define PIXEL_VC(n)      (n << 6)
+#define PIXEL_WC(n)      (n << 8)
+#define PIXEL_ECC(n)     (n << 24)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:169]:
```suggestion
#define PIXEL_DT(n)      ((n) << 0)
#define PIXEL_VC(n)      ((n) << 6)
#define PIXEL_WC(n)      ((n) << 8)
#define PIXEL_ECC(n)     ((n) << 24)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 11]
```diff
@@ -0,0 +1,993 @@
...
+/* DSI Pixel Package Register 2 (A31 Page 849) */
+
+#define DSI_PIXEL_PF0_REG (A64_DSI_ADDR + 0x98)
+#define CRC_FORCE         0xffff
+
+/* DSI Pixel Package Register 3 (A31 Page 849) */
+
+#define DSI_PIXEL_PF1_REG (A64_DSI_ADDR + 0x9c)
+#define CRC_INIT_LINE0(n) (n << 0)
+#define CRC_INIT_LINEN(n) (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:180]:
```suggestion
#define CRC_INIT_LINE0(n) ((n) << 0)
#define CRC_INIT_LINEN(n) ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 12]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Sync Package Register 0 (A31 Page 850) */
+
+#define DSI_SYNC_HSS_REG (A64_DSI_ADDR + 0xb0)
+#define SYNC_ECC(n)      (n << 24)
+#define SYNC_D1(n)       (n << 16)
+#define SYNC_D0(n)       (n << 8)
+#define SYNC_VC(n)       (n << 6)
+#define SYNC_DT(n)       (n << 0)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:189]:
```suggestion
#define SYNC_ECC(n)      ((n) << 24)
#define SYNC_D1(n)       ((n) << 16)
#define SYNC_D0(n)       ((n) << 8)
#define SYNC_VC(n)       ((n) << 6)
#define SYNC_DT(n)       ((n) << 0)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 13]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Blank Package Register 2 (A31 Page 852) */
+
+#define DSI_BLK_HBP0_REG (A64_DSI_ADDR + 0xc8)
+
+/* DSI Blank Package Register 3 (A31 Page 852) */
+
+#define DSI_BLK_HBP1_REG (A64_DSI_ADDR + 0xcc)
+#define HBP_PD(n)        (n << 0)
+#define HBP_PF(n)        (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:221]:
```suggestion
#define HBP_PD(n)        ((n) << 0)
#define HBP_PF(n)        ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 14]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Blank Package Register 0 (A31 Page 852) */
+
+#define DSI_BLK_HSA0_REG (A64_DSI_ADDR + 0xc0)
+
+/* DSI Blank Package Register 1 (A31 Page 852) */
+
+#define DSI_BLK_HSA1_REG (A64_DSI_ADDR + 0xc4)
+#define HSA_PD(n)        (n << 0)
+#define HSA_PF(n)        (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:211]:
```suggestion
#define HSA_PD(n)        ((n) << 0)
#define HSA_PF(n)        ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 15]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Blank Package Register 4 (A31 Page 852) */
+
+#define DSI_BLK_HFP0_REG (A64_DSI_ADDR + 0xd0)
+
+/* DSI Blank Package Register 5 (A31 Page 853) */
+
+#define DSI_BLK_HFP1_REG (A64_DSI_ADDR + 0xd4)
+#define HFP_PD(n)        (n << 0)
+#define HFP_PF(n)        (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:231]:
```suggestion
#define HFP_PD(n)        ((n) << 0)
#define HFP_PF(n)        ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 16]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Blank Package Register 6 (A31 Page 853) */
+
+#define DSI_BLK_HBLK0_REG (A64_DSI_ADDR + 0xe0)
+
+/* DSI Blank Package Register 7 (A31 Page 853) */
+
+#define DSI_BLK_HBLK1_REG (A64_DSI_ADDR + 0xe4)
+#define HBLK_PD(n)        (n << 0)
+#define HBLK_PF(n)        (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:241]:
```suggestion
#define HBLK_PD(n)        ((n) << 0)
#define HBLK_PF(n)        ((n) << 16)
```

File: arch/arm64/src/a64/a64_mipi_dsi.c [hunk 17]
```diff
@@ -0,0 +1,993 @@
...
+
+/* DSI Blank Package Register 8 (A31 Page 854) */
+
+#define DSI_BLK_VBLK0_REG (A64_DSI_ADDR + 0xe8)
+
+/* DSI Blank Package Register 9 (A31 Page 854) */
+
+#define DSI_BLK_VBLK1_REG (A64_DSI_ADDR + 0xec)
+#define VBLK_PD(n)        (n << 0)
+#define VBLK_PF(n)        (n << 16)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/a64_mipi_dsi.c:251]:
```suggestion
#define VBLK_PD(n)        ((n) << 0)
#define VBLK_PF(n)        ((n) << 16)
```

File: arch/arm64/src/a64/mipi_dsi.c [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+ * Input Parameters:
+ *   data - Data buffer
+ *   len  - Length of data buffer
+ *
+ * Returned Value:
+ *   MIPI DSI CRC value of the data buffer
+ *
+ ****************************************************************************/
+
+static uint16_t compute_crc(FAR const uint8_t *data, size_t len)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.c:61]:
```suggestion
static uint16_t compute_crc(const uint8_t *data, size_t len)
```

File: arch/arm64/src/a64/mipi_dsi.c [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+ *   di_wc - Packet Header (Data Identifier + Word Count)
+ *   len   - Length of Packet Header (Should be 3 bytes)
+ *
+ * Returned Value:
+ *   MIPI DSI ECC value of the Packet Header
+ *
+ ****************************************************************************/
+
+static uint8_t compute_ecc(FAR const uint8_t *di_wc,
+                           size_t len)
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.c:91]:
```suggestion
static uint8_t compute_ecc(const uint8_t *di_wc, size_t len)
```

File: arch/arm64/src/a64/mipi_dsi.c [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+ *   txbuf   - Payload data for the packet
+ *   txlen   - Length of payload data (Max 65541 bytes)
+ *
+ * Returned Value:
+ *   Number of bytes in the returned packet; ERROR (-1) if packet buffer is
+ *   too small for the packet
+ *
+ ****************************************************************************/
+
+ssize_t mipi_dsi_long_packet(FAR uint8_t *pktbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.c:171]:
```suggestion
ssize_t mipi_dsi_long_packet(uint8_t *pktbuf,
```

File: arch/arm64/src/a64/mipi_dsi.c [hunk 21]
```diff
@@ -0,0 +1,387 @@
...
+                             size_t pktlen,
+                             uint8_t channel,
+                             enum mipi_dsi_e cmd,
+                             FAR const uint8_t *txbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.c:175]:
```suggestion
                             const uint8_t *txbuf,
```

File: arch/arm64/src/a64/mipi_dsi.c [hunk 22]
```diff
@@ -0,0 +1,387 @@
...
+ *   txbuf   - Payload data for the packet
+ *   txlen   - Length of payload data (1 or 2 bytes)
+ *
+ * Returned Value:
+ *   Number of bytes in the returned packet; ERROR (-1) if packet buffer is
+ *   too small for the packet
+ *
+ ****************************************************************************/
+
+ssize_t mipi_dsi_short_packet(FAR uint8_t *pktbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.c:299]:
```suggestion
ssize_t mipi_dsi_short_packet(uint8_t *pktbuf,
```

File: arch/arm64/src/a64/mipi_dsi.c [hunk 23]
```diff
@@ -0,0 +1,387 @@
...
+                              size_t pktlen,
+                              uint8_t channel,
+                              enum mipi_dsi_e cmd,
+                              FAR const uint8_t *txbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.c:303]:
```suggestion
                              const uint8_t *txbuf,
```

File: arch/arm64/src/a64/mipi_dsi.h [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+ *   txbuf   - Payload data for the packet
+ *   txlen   - Length of payload data (Max 65541 bytes)
+ *
+ * Returned Value:
+ *   Number of bytes in the returned packet; ERROR (-1) if packet buffer is
+ *   too small for the packet
+ *
+ ****************************************************************************/
+
+ssize_t mipi_dsi_long_packet(FAR uint8_t *pktbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.h:72]:
```suggestion
ssize_t mipi_dsi_long_packet(uint8_t *pktbuf,
```

File: arch/arm64/src/a64/mipi_dsi.h [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+                             size_t pktlen,
+                             uint8_t channel,
+                             enum mipi_dsi_e cmd,
+                             FAR const uint8_t *txbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.h:76]:
```suggestion
                             const uint8_t *txbuf,
```

File: arch/arm64/src/a64/mipi_dsi.h [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+ *   txbuf   - Payload data for the packet
+ *   txlen   - Length of payload data (1 or 2 bytes)
+ *
+ * Returned Value:
+ *   Number of bytes in the returned packet; ERROR (-1) if packet buffer is
+ *   too small for the packet
+ *
+ ****************************************************************************/
+
+ssize_t mipi_dsi_short_packet(FAR uint8_t *pktbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.h:101]:
```suggestion
ssize_t mipi_dsi_short_packet(uint8_t *pktbuf,
```

File: arch/arm64/src/a64/mipi_dsi.h [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+                              size_t pktlen,
+                              uint8_t channel,
+                              enum mipi_dsi_e cmd,
+                              FAR const uint8_t *txbuf,
```
Comment (pkarashchenko) [inline arch/arm64/src/a64/mipi_dsi.h:105]:
```suggestion
                              const uint8_t *txbuf,
```

**Comment by lupyuen** [comment]
I just rebased with master, hope this fixes the macOS CI

**Pull request merged**